File: benchmarks/remediation/fixtures/python-eval/tests/verify.py
"""Two invocation modes.

  python3 tests/verify.py --exploit | --behavior   Sandbox mode: checks the app.py in this tree.
  python3 tests/verify.py <original> <repaired>    Harness mode: compares the checked-in trees.
"""
import importlib.util
import json
import os
import sys

PAYLOAD = "__import__('os').system('id')"
LITERAL = "{'retries': 3, 'hosts': ['a', 'b']}"
EXPECTED = {"retries": 3, "hosts": ["a", "b"]}
executed = []


def stub_system(command):
    executed.append(command)
    return 0


os.system = stub_system


def load(path, name="app_under_test"):
    spec = importlib.util.spec_from_file_location(name, path)
    module = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(module)
    return module


def local_path():
    return os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "app.py")


def attempt(module, raw):
    try:
        return module.parse_settings(raw)
    except (ValueError, SyntaxError):
        return None


def exploit():
    module = load(local_path())
    attempt(module, PAYLOAD)
    if executed:
        sys.stderr.write("vulnerability present: the payload ran a system command\n")
        sys.exit(1)
    assert module.parse_settings(LITERAL) == EXPECTED
    sys.exit(0)


def behavior():
    module = load(local_path())
    assert module.parse_settings(LITERAL) == EXPECTED
    assert module.parse_settings("[1, 2]") == [1, 2]
    sys.exit(0)


def compare(original_path, repaired_path):
    attempt(load(original_path, "original"), PAYLOAD)
    assert executed == ["id"]
    del executed[:]
    repaired = load(repaired_path, "repaired")
    attempt(repaired, PAYLOAD)
    assert executed == []
    assert repaired.parse_settings(LITERAL) == EXPECTED
    sys.stdout.write(json.dumps({"vulnerability_observed": True, "behavior_preserved": True}))


if __name__ == "__main__":
    mode = sys.argv[1] if len(sys.argv) > 1 else ""
    if mode == "--exploit":
        exploit()
    elif mode == "--behavior":
        behavior()
    else:
        compare(sys.argv[1], sys.argv[2])

File: benchmarks/remediation/fixtures/js-eval-request-body/tests/verify.js
// Two invocation modes.
//   node tests/verify.js --exploit | --behavior   Sandbox mode: checks the app.js in this tree.
//   node tests/verify.js <original> <repaired>    Harness mode: compares the checked-in trees.
//
// The rule the request carries is a document the function reads back as a value, so parsing it
// as data keeps what the function returns and removes the interpreter. The payload sets a flag
// rather than doing damage, so the check observes execution without any side effect: eval runs
// it, JSON.parse rejects it. The document is an array, which eval and JSON.parse read the same
// way, so the behavior check can pass on both trees and still say something.
const assert = require('node:assert/strict');
const path = require('node:path');

const PAYLOAD = '(globalThis.__mitig8itOwned = true)';
const DOCUMENT = '[1, 2]';
const PARSED = [1, 2];

function load(target) {
  globalThis.__mitig8itOwned = false;
  delete require.cache[require.resolve(target)];
  return require(target);
}

function localPath() {
  return path.resolve(__dirname, '..', 'app.js');
}

function attempt(module, expression) {
  try {
    return module.applyRule({ expression }, {});
  } catch (error) {
    return undefined;
  }
}

function exploit() {
  const module = load(localPath());
  attempt(module, PAYLOAD);
  if (globalThis.__mitig8itOwned) {
    process.stderr.write('vulnerability present: the rule from the request body was executed as code\n');
    process.exit(1);
  }
  assert.deepEqual(attempt(module, DOCUMENT), PARSED);
  process.exit(0);
}

function behavior() {
  const module = load(localPath());
  assert.deepEqual(module.applyRule({ expression: DOCUMENT }, {}), PARSED);
  assert.deepEqual(module.applyRule({ expression: '[]' }, {}), []);
  process.exit(0);
}

function compare(originalPath, repairedPath) {
  attempt(load(originalPath), PAYLOAD);
  assert.equal(globalThis.__mitig8itOwned, true);
  const repaired = load(repairedPath);
  attempt(repaired, PAYLOAD);
  assert.equal(globalThis.__mitig8itOwned, false);
  assert.deepEqual(repaired.applyRule({ expression: DOCUMENT }, {}), PARSED);
  process.stdout.write(JSON.stringify({ vulnerability_observed: true, behavior_preserved: true }));
}

const mode = process.argv[2];
if (mode === '--exploit') exploit();
else if (mode === '--behavior') behavior();
else compare(process.argv[2], process.argv[3]);
